From 093a73f4229f411f93b8a5ead748eec5e9fb58e0 Mon Sep 17 00:00:00 2001
From: Andreas Schultz <andreas.schultz@travelping.com>
Date: Tue, 27 Apr 2021 18:51:41 +0200
Subject: [PATCH 1/3] diameter: fix set unodered flag in client (connect) mode

The semantics of switching a SCTP from ordered to unorderd are not
documented anywhere. The code as using assoc_id == 0, which mean
"all future associations". Adding a stream to the current association
is by strict reading not a new association, but an extension to the
current existing one. None the less, this work on the server
(listening) side.

It does not work on the client (connecting) side, at least on recent
Linux kernels. Using a value of 1 (SCTP_CURRENT_ASSOC) makes it work
on both sides.
---
 lib/diameter/src/transport/diameter_sctp.erl | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/diameter/src/transport/diameter_sctp.erl b/lib/diameter/src/transport/diameter_sctp.erl
index b6a25a9a63..77feb966a9 100644
--- a/lib/diameter/src/transport/diameter_sctp.erl
+++ b/lib/diameter/src/transport/diameter_sctp.erl
@@ -826,8 +826,9 @@ recv(#transport{rotate = N} = S) ->
 unordered(Sock, OS, B)
   when B;
        is_integer(B), OS =< B ->
+    %% assoc_id == SCTP_CURRENT_ASSOC (1)
     inet:setopts(Sock, [{sctp_default_send_param,
-                         #sctp_sndrcvinfo{flags = [unordered]}}]);
+                         #sctp_sndrcvinfo{flags = [unordered], assoc_id = 1}}]);
 
 unordered(_, OS, B)
   when not B;
-- 
2.30.2

